optimize pytorch layer_norm on CPU

xiaomengy opened the pull request:
Summary:
optimize pytorch layer_norm on CPU
both forward and backward will be 4x faster than before

Differential Revision: D15194600

cpuhrsch reviewed (changes requested):
Please move this under [naitive/cpu](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/cpu) to take advantage of the avx dispatch and verify performance in open source.

cpuhrsch commented:
cc @VitalyFedyunin

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1363,6 +1363,18 @@
 
+- func: native_layer_norm(Tensor input, Tensor? weight, Tensor? bias, int[] normalized_shape, float eps) -> (Tensor, Tensor, Tensor)
+  dispatch:
+    CPU: LayerNormForwardCPU
```
can you use the standard camel case, i.e. `layer_norm_forward_cpu` throughout this PR?

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1363,6 +1363,18 @@
     CPU: batch_norm_update_stats_cpu
     CUDA: batch_norm_update_stats_cuda
 
+- func: native_layer_norm(Tensor input, Tensor? weight, Tensor? bias, int[] normalized_shape, float eps) -> (Tensor, Tensor, Tensor)
```
can you give these outputs names, e.g. `(Tensor output, Tensor mean, Tensor rstd)` or similar?

BIT-silence replied:
Done.

BIT-silence replied:
It seems this change will break the yaml test. I got "only whitelisted operators are allowed to have named return type, got native_layer_norm". Just removed the return tensor names.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1363,6 +1363,18 @@
+  dispatch:
+    CPU: LayerNormBackwardCPU
+
+- func: native_layer_norm_double_backward(Tensor? ggI, Tensor? ggW, Tensor? ggb, Tensor gO, Tensor input, Tensor mean, Tensor rstd, Tensor? weight, int[] normalized_shape, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
```
same with the output names here.

VitalyFedyunin commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -466,6 +468,15 @@ Tensor layer_norm(const Tensor& input, IntArrayRef normalized_shape,
     const Tensor& weight /* optional */, const Tensor& bias /* optional */,
     double eps, bool cudnn_enabled) {
 
+    if (!input.is_cuda()) {
```
input.device().is_cpu() as we are also support HIP and potentially other tensors

BIT-silence replied:
Thanks for the advice!

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1363,6 +1363,18 @@
+  dispatch:
+    CPU: LayerNormForwardCPU
+
+- func: native_layer_norm_backward(Tensor grad_out, Tensor input, Tensor mean, Tensor rstd, Tensor? weight, int[] normalized_shape, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
```
same with output names here.

VitalyFedyunin commented:
Can you please add motivation about returning tuple of 3 elements in compare to original layer_norm?

gchanan commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -466,6 +468,15 @@ Tensor layer_norm(const Tensor& input, IntArrayRef normalized_shape,
     const Tensor& weight /* optional */, const Tensor& bias /* optional */,
     double eps, bool cudnn_enabled) {
 
+    if (!input.is_cuda()) {
```
hmm, I guess we don't have `is_cpu` :(.  I'd make this `input.device().type() == kCPU` or similar since we have other devices now.

VitalyFedyunin replied:
input.device().is_cpu()

BIT-silence replied:
Done

cpuhrsch commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -0,0 +1,647 @@
+#include <ATen/NativeFunctions.h>
```
A lot of this stuff can be rewritten using TensorIterator or existing abstractions such as "mean_std". We need to sit down and write up more docs on those.

fmassa replied:
Here is the `mean_std` PR with implementation for reference https://github.com/pytorch/pytorch/pull/18731

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1363,6 +1363,18 @@
+  dispatch:
+    CPU: LayerNormBackwardCPU
+
+- func: native_layer_norm_double_backward(Tensor? ggI, Tensor? ggW, Tensor? ggb, Tensor gO, Tensor input, Tensor mean, Tensor rstd, Tensor? weight, int[] normalized_shape, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
```
What does the "native_" prefix mean?

BIT-silence replied:
Actually I am not quite sure about that, just follow the native_batch_norm function style. Is it better to remove it?

gchanan commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -0,0 +1,647 @@
+    const IntArrayRef normalized_shape,
+    const double eps) {
+  const int normalized_ndim = normalized_shape.size();
+  AT_CHECK(
```
wait, this is the same error checking code as above, but it just uses different variable names?  That is just unnecessarily confusing, no?

xiaomengy committed:
optimize pytorch layer_norm on CPU (#20197)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20197

optimize pytorch layer_norm on CPU
both forward and backward will be 4x faster than before

Differential Revision: D15194600

fbshipit-source-id: 9cda86f73c31b4b7c50ba370735f6812a2e89dc2
